File: Tutorials/Scala/5 Topics/consumer/src/main/scala/ru/example/rmq/topics/Consumer.scala
package ru.example.rmq.topics

import com.rabbitmq.client._
import java.nio.charset.StandardCharsets

object Consumer {
  def main(args: Array[String]): Unit = {
    val EXCHANGE_NAME = "topic_logs"
    val virtualHost   = "/"

    if (args.length < 4) {
      println("Usage: Consumer hostname username password binding_key [binding_key...]")
      sys.exit(-1)
    }

    val hostname = args(0)
    val username = args(1)
    val password = args(2)

    val factory = new ConnectionFactory
    factory.setHost(hostname)
    factory.setUsername(username)
    factory.setPassword(password)
    factory.setVirtualHost(virtualHost)

    try {
      val connection = factory.newConnection
      val channel    = connection.createChannel

      channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.TOPIC)
      val queueName = channel.queueDeclare().getQueue

      args.slice(3, args.length).foreach { bindingKey =>
        println(s"Bind to $bindingKey")
        channel.queueBind(queueName, EXCHANGE_NAME, bindingKey)
      }

      println("[*] Waiting for messages. To exit press CTRL+C")
      channel.basicConsume(
        queueName,
        true,
        (_: String, delivery: Delivery) => {
          val message = new String(delivery.getBody, StandardCharsets.UTF_8)
          println(s"[x] Received '${delivery.getEnvelope.getRoutingKey}': '$message'")
        },
        (_: String, _: ShutdownSignalException) => {}
      )
    } catch {
      case e: Exception =>
        println(e.getLocalizedMessage)
        sys.exit(-1)
    }
  }
}

File: Tutorials/Java/5 Topics/Consumer/src/main/java/Consumer.java
import com.rabbitmq.client.*;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

public class Consumer {
    public static void main(String[] argv) throws Exception {
        String hostname;
        String username;
        String password;
        String EXCHANGE_NAME = "topic_logs";
        String virtualHost = "/";

        if (argv.length < 4) {
            System.err.println("Usage: Consumer host username password binding_key [binding_key...]");
            System.exit(1);
        }

        hostname = argv[0];
        username = argv[1];
        password = argv[2];

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(hostname);
        factory.setUsername(username);
        factory.setPassword(password);
        factory.setVirtualHost(virtualHost);

        try (Connection connection = factory.newConnection();
             Channel channel = connection.createChannel()) {

            channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.TOPIC);
            String queueName = channel.queueDeclare().getQueue();

            for (String bindingKey : Arrays.copyOfRange(argv, 3, argv.length)) {
                channel.queueBind(queueName, EXCHANGE_NAME, bindingKey);
            }
            System.out.println(" [*] Waiting for messages. To exit press CTRL+C");

            DeliverCallback deliverCallback = (consumerTag, delivery) -> {
                String message = new String(delivery.getBody(), StandardCharsets.UTF_8);
                System.out.println(" [x] Received '" + delivery.getEnvelope().getRoutingKey() + "':'" + message + "'");
            };

            while(true)
                channel.basicConsume(queueName, true, deliverCallback, consumerTag -> {
            });
        }
    }
}